File: leetcode/C++/unique-paths-iii.cpp
class Solution {
public:
    int dy[4]={1,-1,0,0}, dx[4]={0,0,1,-1},ny,nx;
    int sy,sx,ey,ex,count;
    vector<vector<int>> visit;
    int traverse(vector<vector<int>>& grid, int y, int x, int count){
        if(count==0) return y==sy && x==sx;
        int res=0;
        visit[y][x]=1;
        for(int i=0; i<4; ++i){
            ny=y+dy[i]; nx=x+dx[i];
            if(ny<0 || ny>=grid.size() || nx<0 || nx>=grid[0].size() || grid[ny][nx]==-1 || visit[ny][nx]) continue;
            res+=traverse(grid,ny,nx,count-1);
        }
        visit[y][x]=0;
        return res;
    }
    int uniquePathsIII(vector<vector<int>>& grid) {
        sy=sx=ey=ex=count=0;
        visit=vector<vector<int>>(grid.size(), vector<int>(grid[0].size(),0));
        for(int i=0; i<grid.size(); ++i){
            for(int j=0; j<grid[i].size(); ++j){
                if(grid[i][j]!=-1) ++count; 
                if(grid[i][j]==1) {sy=i; sx=j;}
                else if(grid[i][j]==2) {ey=i; ex=j;}
            }
        }
        return traverse(grid, ey, ex, count-1);
    }
};

File: leetcode/Python/unique-paths-iii.py
class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        sy,sx,ey,ex,dy,dx, c = 0, 0, 0, 0, [1,-1,0,0], [0,0,1,-1], 0
        visit = [[0 for _ in range(len(grid[0]))] for _ in range(len(grid))]
        def traverse(y,x,cnt):
            if cnt==0:
                return 1 if y==sy and x==sx else 0
            visit[y][x]=1
            res, ny , nx = 0,0,0
            for i in range(4):
                ny, nx = y+dy[i], x+dx[i]
                if ny<0 or ny>=len(grid) or nx<0 or nx>=len(grid[0]) or grid[ny][nx]==-1 or visit[ny][nx]==1:
                    continue
                res += traverse(ny,nx,cnt-1)
            visit[y][x]=0
            return res
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j]!=-1:
                    c+=1
                if grid[i][j]==1:
                    sy, sx = i, j
                elif grid[i][j]==2:
                    ey, ex = i, j
        return traverse(ey,ex,c-1)